pymeasure/instruments/agilent/agilent33521A.py

```diff
@@ -0,0 +1,442 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2019 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+# Parts of this code were copied and adapted from the Agilent33220A class.
+
+import logging
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import strict_discrete_set,\
+    strict_range, joined_validators
+from time import time
+from pyvisa.errors import VisaIOError
+
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+# Capitalize string arguments to allow for better conformity with other WFG's
+def capitalize_string(string: str, *args, **kwargs):
+    return string.upper()
+
+
+# Combine the capitalize function and validator
+string_validator = joined_validators(capitalize_string, strict_discrete_set)
```

Comment:
I would be happy to do so, but just for the record, the validators were not done by myself, but by the author of the Agilent33220A class.

Reply:
Thanks for pointing that out. It will be nice to centralize the validator. Can you update the `Agilent33220A` class to rely on the common validator?

Reply:
Yeah, I will do it this weekend.

Reply:
Ok, correct me if I am wrong, but I believe this string_validator function is identical to capitalize string. The joined_validator function definition uses this statement:

`for validator, vals in zip(validators, values):`

The zip command expects a list for both validators and values and will reduce both lists to the smallest length.
The string_validator function is defined as:

`string_validator = joined_validators(capitalize_string, strict_discrete_set)`

and is always called with only one single values list. The idea is, that the value given is capitalized and then checked against the values range provided. Problem is, that since only one values range is provided, the zip statement removes the second function given in
'joined_validators(capitalize_string, strict_discrete_set)`.

A simple check:

`test_set = ['TEST']`
`strict_discrete_set('TEST', test_set)` --> this will run without error
`string_validator('TEST', test_set)` --> this will run without error
`string_validator('TOAST', test_set)` --> this will also run without error. This should produce an error. What it does instead is to return "TOAST".
`string_validator('toast', test_set)` also returns "TOAST".

I think this was easy to miss, since it does not create an error at runtime and it capitalizes the values correctly. I am not sure, this capitalization is necessary (it would be kind of nice though)...